File: Simulator/Effects/Effect.cs
using System.Diagnostics;
using System.Windows.Media;

namespace Simulator.Effects
{
    public abstract class Effect
    {
        private long lastMillis;
        protected readonly Matrix matrix;
        protected readonly Stopwatch watch = Stopwatch.StartNew();

        public Color Background { get; set; } = Colors.Black;

        protected Effect(Matrix matrix)
        {
            this.matrix = matrix;
        }

        public void Setup()
        {
            Clear();
            lastMillis = 0;
            watch.Restart();
            SetupInternal();
        }

        protected virtual void SetupInternal()
        {
        }

        public void Loop()
        {
            var ms = watch.ElapsedMilliseconds;
            var dt = ms - lastMillis;
            lastMillis = ms;

            if (dt > 0)
                LoopInternal(dt);
        }

        protected virtual void LoopInternal(long dt)
        {
        }

        protected void Clear()
        {
            foreach (var pixel in matrix.Pixels)
                pixel.Color = Background;
        }
    }
}

File: NyMatrix/Effect.cpp
#include "Effect.h"

void Effect::setup()
{
    clear();
    lastMillis = 0;
    setupInternal();
}

void Effect::loop()
{
    auto ms = millis();
    auto dt = ms - lastMillis;
    lastMillis = ms;

    if (dt > 0)
        loopInternal(dt);
}

void Effect::clear()
{
    for (int x = 0; x < MATRIX_WIDTH; x++)
        for (int y = 0; y < MATRIX_WIDTH; y++)
            matrix->SetPixelColor(topo->Map(x, y), background);
}

void Effect::dim()
{
    auto dimRatio = (byte)(brightness * 255);
    if (dimRatio < 255)
    {
        for (int x = 0; x < MATRIX_WIDTH; x++)
        {
            for (int y = 0; y < MATRIX_WIDTH; y++)
            {
                auto color = matrix->GetPixelColor(topo->Map(x, y));
                matrix->SetPixelColor(topo->Map(x, y), color.Dim(dimRatio));
            }
        }
    }
}
